### ISSUE TITLE
implement flux hostlist command

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#3344]
discussed several times in other issues.  this could be convenient for scripting, should be done after PR #3341 is done.

**Comment by grondo** [comment]
We also had a `flux hostlist` command in the days of wreckrun which had a different kind of use case. We should look back at that command and see if we can usefully combine the two use cases.

**Comment by chu11** [comment]
I thought of cranking this one out only to realized I would just be re-implementing https://github.com/grondo/lua-hostlist

perhaps could be copied in and tweaked?

**Comment by grondo** [comment]
This one has languished since there isn't really a current use case.

However, some local issues have been opened up against `lua-hostlist` that could be solved by implementing a `flux hostlist`. We can't just copy it in an tweak it because `libflux-hostlist` API and the original hostlist API are quite different.

Let's pause on this one until I decide if it is easier to fix the issues with `lua-hostlist` by updating that code, or whether we should move the code here.

**Comment by grondo** [comment]
A user asked how to create a hostlist file for a job and I realized there isn't an easy way at the moment (without resorting to `lua-hostlist`).

I've prototyped a simple `flux-hostlist` command which takes an option set of "sources", including:
 - "local" - get a hostlist from the enclosing job or instance
 - jobid - get hostlist for named jobid
 - "stdin" or "-" - get a hostlist from stdin
 - raw hosts - treat argument as hostlist

The default would be "local", so that `flux hostlist` gives you a hostlist for the enclosing instance/job:
```console
$ src/cmd/flux hostlist
pi[3,0-2,4]
```

But you can also target jobids so you can do things like find the common hosts for the last 5 failed jobs:
```console
$ src/cmd/flux hostlist --intersect $(flux jobs -f failed -c 5  -no {id})
pi[0,3]
```

Otherwise, the command could be used similarly to lua-hostlist (with more limited options for now)
```
$ src/cmd/flux hostlist --help
usage: flux-hostlist [-h] [-e] [-d S] [-c] [-n NTH] [-L LIMIT] [-S] [-u] [-i]
                     [--fallback]
                     [SOURCES ...]

positional arguments:
  SOURCES            One or more hostlist sources: 'local' (job or current
                     instance), '-' or 'stdin' (read hosts on stdin), 'jobid',
                     or 'hosts' (default: local)

optional arguments:
  -h, --help         show this help message and exit
  -e, --expand       Expand hostlist using defined output delimiter
  -d, --delimiter=S  Set output delimiter for expanded hostlist (default=',')
  -c, --count        Print the total number of hosts
  -n, --nth=NTH      Output host at index N (-N to index from end)
  -L, --limit=LIMIT  Output at most N hosts (-N for the last N hosts)
  -S, --sort         Return sorted result
  -u, --union        Return the union of all hostlists. Default is to append
  -i, --intersect    Return the intersection of all hostlists
      --fallback     Fallback to treating jobids that are not found as
                     hostnames (for hostnames that are also valid jobids e.g.
                     f1, fuzz100, etc)
```

**Comment by garlick** [comment]
Neat!  Heh, if I wanted to know what node rank 7 of my last job ran on I could do:
```
jobid=$(flux job last)
flux hostlist --nth $(flux taskmap --nodeid 7 $jobid) 
```

**Comment by grondo** [comment]
Should have been closed by #5724
